File: DynamicKey/AgoraDynamicKey/rust/src/apaas_token_builder.rs
use crate::access_token;
use crate::utils;

// Build room user token.
//
// app_id: The App ID issued to you by Agora. Apply for a new App ID from Agora Dashboard if it is missing from your kit. See Get an App ID.
// app_certificate: Certificate of the application that you registered in the Agora Dashboard. See Get an App Certificate.
// room_uuid: The room's id, must be unique.
// user_uuid: The user's id, must be unique.
// role: The user's role.
// expire: represented by the number of seconds elapsed since 1/1/1970. If, for example, you want to access the Agora Service within 10 minutes after the token is
//    generated, set expireTimestamp as the current timestamp + 600 (seconds).
// return The room user token.
pub fn build_room_user_token(
    app_id: &str, app_certificate: &str, room_uuid: &str, user_uuid: &str, role: i16, expire: u32,
) -> Result<String, Box<dyn std::error::Error>> {
    let mut token = access_token::new_access_token(app_id, app_certificate, expire);

    let chat_user_id = utils::md5(user_uuid);
    let mut service_apaas = access_token::new_service_apaas(room_uuid, user_uuid, role);
    service_apaas.service.add_privilege(access_token::PRIVILEGE_APAAS_ROOM_USER, expire);
    token.add_service(Box::new(service_apaas));

    let mut service_rtm = access_token::new_service_rtm(user_uuid);
    service_rtm.service.add_privilege(access_token::PRIVILEGE_LOGIN, expire);
    token.add_service(Box::new(service_rtm));

    let mut service_chat = access_token::new_service_chat(&chat_user_id);
    service_chat.service.add_privilege(access_token::PRIVILEGE_CHAT_USER, expire);
    token.add_service(Box::new(service_chat));

    return token.build();
}

// Build user token.
//
// app_id: The App ID issued to you by Agora. Apply for a new App ID from Agora Dashboard if it is missing from your kit. See Get an App ID.
// app_certificate: Certificate of the application that you registered in the Agora Dashboard. See Get an App Certificate.
// user_uuid: The user's id, must be unique.
// expire: represented by the number of seconds elapsed since 1/1/1970. If, for example, you want to access the Agora Service within 10 minutes after the token is
//    generated, set expireTimestamp as the current timestamp + 600 (seconds).
// return The user token.
pub fn build_user_token(app_id: &str, app_certificate: &str, user_uuid: &str, expire: u32) -> Result<String, Box<dyn std::error::Error>> {
    let mut token = access_token::new_access_token(app_id, app_certificate, expire);

    let mut service_apaas = access_token::new_service_apaas("", user_uuid, -1);
    service_apaas.service.add_privilege(access_token::PRIVILEGE_APAAS_USER, expire);
    token.add_service(Box::new(service_apaas));

    return token.build();
}

// Build app token.
//
// app_id: The App ID issued to you by Agora. Apply for a new App ID from Agora Dashboard if it is missing from your kit. See Get an App ID.
// app_certificate: Certificate of the application that you registered in the Agora Dashboard. See Get an App Certificate.
// expire: represented by the number of seconds elapsed since 1/1/1970. If, for example, you want to access the Agora Service within 10 minutes after the token is
//    generated, set expireTimestamp as the current timestamp + 600 (seconds).
// return The app token.
pub fn build_app_token(app_id: &str, app_certificate: &str, expire: u32) -> Result<String, Box<dyn std::error::Error>> {
    let mut token = access_token::new_access_token(app_id, app_certificate, expire);

    let mut service_apaas = access_token::new_service_apaas("", "", -1);
    service_apaas.service.add_privilege(access_token::PRIVILEGE_APAAS_APP, expire);
    token.add_service(Box::new(service_apaas));

    return token.build();
}

File: DynamicKey/AgoraDynamicKey/php/src/ApaasTokenBuilder.php
<?php

require_once "AccessToken2.php";

class ApaasTokenBuilder
{
    /**
     * Build room user token.
     *
     * @param $appId The App ID issued to you by Agora. Apply for a new App ID from
     *     Agora Dashboard if it is missing from your kit. See Get an App ID.
     * @param $appCertificate Certificate of the application that you registered in
     *     the Agora Dashboard. See Get an App Certificate.
     * @param $roomUuid The room's id, must be unique.
     * @param $userUuid The user's id, must be unique.
     * @param $role The user's role.
     * @param $expire represented by the number of seconds elapsed since now. If, for example, you want to access the
     *     Agora Service within 10 minutes after the token is generated, set expire as 600(seconds).
     * @return The room user token.
     */
    public static function buildRoomUserToken($appId, $appCertificate, $roomUuid, $userUuid, $role, $expire)
    {
        $accessToken = new AccessToken2($appId, $appCertificate, $expire);

        $chatUserId = md5($userUuid);
        $serviceApaas = new ServiceApaas($roomUuid, $userUuid, $role);
        $serviceApaas->addPrivilege($serviceApaas::PRIVILEGE_ROOM_USER, $expire);
        $accessToken->addService($serviceApaas);

        $serviceRtm = new ServiceRtm($userUuid);
        $serviceRtm->addPrivilege($serviceRtm::PRIVILEGE_LOGIN, $expire);
        $accessToken->addService($serviceRtm);

        $serviceChat = new ServiceChat($chatUserId);
        $serviceChat->addPrivilege($serviceChat::PRIVILEGE_USER, $expire);
        $accessToken->addService($serviceChat);

        return $accessToken->build();
    }

    /**
     * Build user token.
     *
     * @param $appId The App ID issued to you by Agora. Apply for a new App ID from
     *     Agora Dashboard if it is missing from your kit. See Get an App ID.
     * @param $appCertificate Certificate of the application that you registered in
     *     the Agora Dashboard. See Get an App Certificate.
     * @param $userUuid The user's id, must be unique.
     * @param $expire represented by the number of seconds elapsed since now. If, for example, you want to access the
     *     Agora Service within 10 minutes after the token is generated, set expire as 600(seconds).
     * @return The user token.
     */
    public static function buildUserToken($appId, $appCertificate, $userUuid, $expire)
    {
        $accessToken = new AccessToken2($appId, $appCertificate, $expire);
        $serviceApaas = new ServiceApaas("", $userUuid);
        $serviceApaas->addPrivilege($serviceApaas::PRIVILEGE_USER, $expire);
        $accessToken->addService($serviceApaas);

        return $accessToken->build();
    }

    /**
     * Build app token.
     *
     * @param $appId The App ID issued to you by Agora. Apply for a new App ID from
     *     Agora Dashboard if it is missing from your kit. See Get an App ID.
     * @param $appCertificate Certificate of the application that you registered in
     *     the Agora Dashboard. See Get an App Certificate.
     * @param $expire represented by the number of seconds elapsed since now. If, for example, you want to access the
     *     Agora Service within 10 minutes after the token is generated, set expire as 600(seconds).
     * @return The app token.
     */
    public static function buildAppToken($appId, $appCertificate, $expire)
    {
        $accessToken = new AccessToken2($appId, $appCertificate, $expire);
        $serviceApaas = new ServiceApaas();
        $serviceApaas->addPrivilege($serviceApaas::PRIVILEGE_APP, $expire);
        $accessToken->addService($serviceApaas);

        return $accessToken->build();
    }
}
